File: C++/프로그래머스/3/43163. 단어 변환/단어 변환.cpp
#include <iostream>
#include <string>
#include <vector>
#include <queue>

using namespace std;

int mismatchedCount(string begin, string cur)
{
    int cnt = 0;
    
    for(int i = 0; i < begin.length(); i++)
    {
        if(begin[i] != cur[i])
        {
            cnt++;
        }
    }
    
    return cnt;
}

int solution(string begin, string target, vector<string> words) {
    int answer = 0;
    int size = words.size();
    
    queue<pair<string, int>> q;
    vector<bool> visited(size, false);
    
    q.push({begin, 0});
    
    while(!q.empty())
    {
        string word = q.front().first;
        int cnt = q.front().second;
        q.pop();
        
        if(word.compare(target) == 0) 
        {
            answer = cnt;
            break;
        }
        
        for(int i = 0; i < size; i++)
        {
            string cur = words[i];
            
            if(!visited[i] && mismatchedCount(word, cur) == 1)
            {
                visited[i] = true;
                q.push({cur, cnt + 1});
            }
        }
    }
    
    return answer;
}

File: Python3/프로그래머스/3/43163. 단어 변환/단어 변환.py
def solution(begin, target, words):
    if target not in words:
        return 0;
    
    queue = [(begin, 0)]
    visited = set();
    visited.add(begin)
    
    while len(queue) > 0:
        pair = queue.pop()
        cur = pair[0]
        count = pair[1]
        
        if(cur == target):
            return count;
        
        for word in words:
            if(word not in visited):
                diff_count = sum(1 for a, b in zip(cur, word) if a != b)
                if diff_count == 1:
                    visited.add(word)
                    queue.append((word, count + 1))
                
    return 0